### PR TITLE
Implement tosa lowerings for gather, select and concat ops.

### PR DESCRIPTION
**Pull request by NatashaKnk**
* As MHLO already had an implementation for concatenate and the inputs/outputs are identical, moved the existing MHLO concatenate implementation into core_ops and referenced it from both. 
* The createBroadcastOpIfNeeded was initially implemented to support binary ops only. However, very small changes were required to enable it in supporting broadcasting for more than two inputs. This can be further expanded by allowing callOpBroadcastableConversion to take in more than two inputs, but it may require more substantial changes that might not be worth it for (what is currently) a single op.

### CODE REVIEW

File: docs/tosa-op-coverage.md [hunk 1]
```diff
@@ -31,10 +31,14 @@ The table below shows the supported TOSA ops.
...
 | sub                    | :heavy_check_mark: | |
 | table                  | :heavy_check_mark: | |
+| **Ternary elementwise ops**
+| select                 | :heavy_check_mark: | |
 | **Other ops**
 | argmax                 | :heavy_check_mark: | |
+| concat                 | :heavy_check_mark: | |
 | conv2d                 | :white_check_mark: | Quantization and dilation not supported |
 | depthwise_conv2d       | :white_check_mark: | Quantization and dilation not supported |
+| gather                 | :heavy_check_mark: | |
```
Comment (jpienaar) [inline docs/tosa-op-coverage.md:41]: Nit: gather below fully_connected to keep these in order.

File: lib/Conversion/TosaToEmitC/TosaToEmitC.cpp [hunk 2]
```diff
@@ -118,6 +118,33 @@ class CallOpConversion : public OpConversionPattern<SrcOp> {
...
+  LogicalResult
+  matchAndRewrite(tosa::ConcatOp concatOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    StringRef funcName = "emitc::tosa::concat";
+    StringAttr callee = rewriter.getStringAttr(funcName);
+
+    ArrayAttr args;
+    ArrayAttr templateArgs =
+        rewriter.getArrayAttr({rewriter.getI64IntegerAttr(concatOp.getAxis()),
```
Comment (jpienaar) [inline lib/Conversion/TosaToEmitC/TosaToEmitC.cpp:137]: concatOp.getAxisAttr() should already be a I64Attr

File: lib/Conversion/TosaToEmitC/TosaToEmitC.cpp [hunk 3]
```diff
@@ -415,8 +442,9 @@ class RsqrtOpConversion : public OpConversionPattern<tosa::RsqrtOp> {
 };
 
 /// Convert `tosa.fully_connected` into an `emitc.call` operation.
```
Comment (jpienaar) [inline lib/Conversion/TosaToEmitC/TosaToEmitC.cpp:444]: This comment seems off.

File: lib/Conversion/TosaToEmitC/TosaToEmitC.cpp [hunk 4]
```diff
@@ -415,8 +442,9 @@ class RsqrtOpConversion : public OpConversionPattern<tosa::RsqrtOp> {
...
-template <typename SrcOp, typename Adaptor = typename SrcOp::Adaptor>
-SmallVector<Value, 2>
+template <size_t operands, typename SrcOp,
```
Comment (jpienaar) [inline lib/Conversion/TosaToEmitC/TosaToEmitC.cpp:445]: Why do we need to pass operands here in template? SmallVector can be resized and seems like it has to match Adaptor::getNumOperands().

File: lib/Conversion/TosaToEmitC/TosaToEmitC.cpp [hunk 5]
```diff
@@ -118,6 +118,33 @@ class CallOpConversion : public OpConversionPattern<SrcOp> {
...
+
+public:
+  ConcatOpConversion(MLIRContext *ctx) : OpConversionPattern(ctx) {}
+
+private:
+  LogicalResult
+  matchAndRewrite(tosa::ConcatOp concatOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    StringRef funcName = "emitc::tosa::concat";
```
Comment (jpienaar) [inline lib/Conversion/TosaToEmitC/TosaToEmitC.cpp:132]: Consider using StringLiteral.

File: lib/Conversion/TosaToEmitC/TosaToEmitC.cpp [hunk 6]
```diff
@@ -610,6 +641,36 @@ class ArithmeticRightShiftOpConversion
...
+
+    ArrayAttr args;
+    ArrayAttr templateArgs;
+
+    SmallVector<Value, 3> broadcastedOperands =
+        createBroadcastOpIfNeeded<3>(selectOp, adaptor, rewriter);
+
+    rewriter.replaceOpWithNewOp<emitc::CallOp>(
+        selectOp, selectOp.getType(), callee, args, templateArgs,
+        ValueRange({broadcastedOperands[0], broadcastedOperands[1],
```
Comment (jpienaar) [inline lib/Conversion/TosaToEmitC/TosaToEmitC.cpp:667]: Aren't these all of broadcastedOperands? I think a ValueRange should automatically be constructed here.

File: reference-implementation/include/emitc/types.h [hunk 7]
```diff
@@ -292,6 +292,34 @@ inline Dest binary(const SrcLeft &x, const SrcRight &y, BinaryOp &&op) {
...
+inline Dest ternary(const SrcA &a, const SrcB &b, const SrcB &c,
+                    TernaryOp &&op) {
+  Dest d;
+  auto first1 = a.begin(), last1 = a.end();
+  auto first2 = b.begin(), first3 = c.begin();
+  auto result = d.begin();
+
+  while (first1 != last1) {
+    *result = op(*first1, *first2, *first3);
+    result++;
```
Comment (jpienaar) [inline reference-implementation/include/emitc/types.h:315]: Nit: prefer pre-increment unless post-increment behavior is needed.

(almost feels like this could have simpler with llvm::zip to bring all together, especially as you probably need to assert that we haven't hit the end of any of these).

File: lib/Conversion/TosaToEmitC/TosaToEmitC.cpp (hunk 5 above)
Comment (marbre) [reply lib/Conversion/TosaToEmitC/TosaToEmitC.cpp:132]: Since we use `StringRef funcName` at many places in our code you could leave as is. I will refactor the code anyway ad using `StringLiteral` is the better way.

**Comment by marbre** [comment]
As I am already oof this year and thus have only limited time to review, @henri-gruender will assist with reviewing the reference implementation and its unit tests.

**Review by henri-gruender — CHANGES_REQUESTED** [review]
Looks good in general.
For the unit tests we try to have test cases for the possible ranks of the tensors and with float and integer types. I think these tests are still fine for now. `concat` already has good tests from the mhlo implementation.

File: lib/Conversion/TosaToEmitC/TosaToEmitC.cpp (hunk 3 above)
Comment (NatashaKnk) [reply lib/Conversion/TosaToEmitC/TosaToEmitC.cpp]: Oops, I actually fixed it before but it got lost somewhere in the process. Fixed it again :)

File: lib/Conversion/TosaToEmitC/TosaToEmitC.cpp (hunk 4 above)
Comment (NatashaKnk) [reply lib/Conversion/TosaToEmitC/TosaToEmitC.cpp]: Good point, for some reason I had assumed we would need to specify the SmallVector size. Fixed, thanks!

File: reference-implementation/include/emitc/types.h (hunk 7 above)
Comment (NatashaKnk) [reply reference-implementation/include/emitc/types.h]: I was thinking the same thing, but I wasn't entirely sure how to wire llvm::zip into here.

File: reference-implementation/include/emitc/tosa.h [hunk 8]
```diff
@@ -270,7 +270,30 @@ inline Tensor<int32_t, Shape...> table(Tensor<int16_t, Shape...> x,
...
+template <int32_t Dimension, typename Dest, typename... Src>
+inline Dest concat(Src... inputs) {
+  return emitc::concatenate<Dimension, Dest, Src...>(inputs...);
+}
+
+template <int32_t Dimension, typename Dest, typename Src1, typename... Src>
+inline Dest concat(Src1 input1, Src... inputs) {
+  return emitc::concatenate<Dimension, Dest, Src1, Src...>(input1, inputs...);
+}
+
```
[the remark this answers is not in the document]
Comment (NatashaKnk) [reply reference-implementation/include/emitc/tosa.h]: Good point! Thanks.

File: reference-implementation/include/emitc/tosa.h [hunk 9]
```diff
@@ -475,6 +498,33 @@ Dest fully_connected(Src input, Weights weights, Bias bias) {
...
+  static_assert(input.dim(0) == indices.dim(0),
+                "Input and weight batch dimension do not match.");
+  static_assert(input.dim(2) == result.dim(2),
+                "Input and output channel dimension do not match.");
+  static_assert(indices.dim(1) == result.dim(1),
+                "Weight and output index dimension do not match.");
+
+  auto it = result.begin();
+  size_t d0offset = Src::dim(1) * Src::dim(2);
+  for (int64_t i = 0, idx = Idx::size(); i < idx; i++) {
```
[the remark this answers is not in the document]
Comment (NatashaKnk) [reply reference-implementation/include/emitc/tosa.h]: Fixed, thanks!

**Comment by NatashaKnk** [comment]
Added a couple of tests, but not entirely sure what other ones should be added. Since a lot of the ops operate on Tosa_Tensor which doesn't have a specific range of possible ranks (as far as I'm aware), how many potential ranks would be considered sufficient? Similarly, when it comes to types, should I test for each potential bit width, or is an all-encompassing int/float test enough?
Any other testing tips are also appreciated :)

**Comment by henri-gruender** [comment]
> Since a lot of the ops operate on Tosa_Tensor which doesn't have a specific range of possible ranks (as far as I'm aware), how many potential ranks would be considered sufficient?

Right, most common are probably Tensor0D through Tensor4D (which are also the ones defined by name in the Tensor class). So if there is no info about the possible ranks, you could stick to those.

> Similarly, when it comes to types, should I test for each potential bit width, or is an all-encompassing int/float test enough?

In my opinion it is enough, if there is at least one test case for an int of any width and one for float or double.

Overall I think the first five test cases for `tosa::clamp` set a good example for a general function, but it is more important to look for edge cases in the implementation and specification. Hopefully we can add some guidelines for the tests in the near future, but for now the current set of tests in the pull request is good enough as it is.

**Review by henri-gruender — APPROVED** [review]

**Review by marbre — APPROVED** [review]
